### PR TITLE
[SPARK-12667] Remove block manager's internal "external block store" API

### PR DESCRIPTION
**Pull request by rxin**
This pull request removes the external block store API. This is rarely used, and the file system interface is actually a better, more standard way to interact with external storage systems.

There are some other things to remove also, as pointed out by @JoshRosen. We will do those as follow-up pull requests.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 1]
```diff
@@ -314,7 +309,7 @@ private[spark] class BlockManager(
       val memSize = if (memoryStore.contains(blockId)) memoryStore.getSize(blockId) else 0L
       val diskSize = if (diskStore.contains(blockId)) diskStore.getSize(blockId) else 0L
       // Assume that block is not in external block store
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Is this comment out-of-date now that we've removed the external block store API?

**Comment by JoshRosen** [comment]
There's some documentation which needs to be updated:

- Remove `spark.externalBlockStore.urlspark.externalBlockStore.*` configurations from the configuration page.
- Update the documentation for the `OFF_HEAP` storage level.
- There's still occurrences of the string "external block store" in `BlockManager` comments / log messages.
- The UI still contains text which references the external block store.

In addition, there are a few more places where dead code can be deleted:

- In `ExecutorExitCode`, remove `EXTERNAL_BLOCK_STORE_FAILED_TO_INITIALIZE` and `EXTERNAL_BLOCK_STORE_FAILED_TO_CREATE_DIR`.
- The `RDDInfo` `@DeveloperAPI` should be updated to either deprecate the external block store size fields or to remove them.
- Log messages can be updated to omit the external block store size information, since it will always be zero now.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 2]
```diff
@@ -313,8 +308,7 @@ private[spark] class BlockManager(
     blockInfo.asScala.get(blockId).map { info =>
       val memSize = if (memoryStore.contains(blockId)) memoryStore.getSize(blockId) else 0L
       val diskSize = if (diskStore.contains(blockId)) diskStore.getSize(blockId) else 0L
-      // Assume that block is not in external block store
-      BlockStatus(info.level, memSize, diskSize, 0L)
+      BlockStatus(info.level, memSize = memSize, diskSize = diskSize)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: why explicitly specify the fields here? type safety?

File: core/src/main/scala/org/apache/spark/storage/StorageLevel.scala [hunk 3]
```diff
@@ -150,7 +150,9 @@ object StorageLevel {
   val MEMORY_AND_DISK_2 = new StorageLevel(true, true, false, true, 2)
   val MEMORY_AND_DISK_SER = new StorageLevel(true, true, false, false)
   val MEMORY_AND_DISK_SER_2 = new StorageLevel(true, true, false, false, 2)
-  val OFF_HEAP = new StorageLevel(false, false, true, false)
+
+  // Redirect to MEMORY_ONLY_SER for now.
+  val OFF_HEAP = MEMORY_ONLY_SER
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/storage/StorageLevel.scala]: should we at least log a warning?

File: core/src/main/scala/org/apache/spark/storage/StorageUtils.scala [hunk 4]
```diff
@@ -48,14 +48,14 @@ class StorageStatus(val blockManagerId: BlockManagerId, val maxMem: Long) {
    * non-RDD blocks for the same reason. In particular, RDD storage information is stored
    * in a map indexed by the RDD ID to the following 4-tuple:
    *
-   *   (memory size, disk size, off-heap size, storage level)
+   *   (memory size, disk size, storage level)
    *
    * We assume that all the blocks that belong to the same RDD have the same storage level.
    * This field is not relevant to non-RDD blocks, however, so the storage information for
    * non-RDD blocks contains only the first 3 fields (in the same order).
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/storage/StorageUtils.scala]: should only contain the first 2 fields

**Comment by andrewor14** [comment]
LGTM, pretty straightforward change. I confirmed that JSON parts are backward compatible.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 2 above)
Comment (rxin) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Yea - it is too easy to swap memSize / diskSize.

**Comment by JoshRosen** [comment]
LGTM as well. I'm going to merge this now in order to unblock progress on other PRs that touch BlockManager. For any smaller changes, such as the doc updates, let's defer them to followups.

**Pull request closed without merging**